// @workflow/typescript-plugin/dist/completions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.enhanceCompletions = enhanceCompletions;
const tsserverlibrary_1 = __importDefault(require("typescript/lib/tsserverlibrary"));
const utils_1 = require("./utils");
function enhanceCompletions(fileName, position, prior, program, ts) {
    if (!prior)
        return prior;
    const sourceFile = program.getSourceFile(fileName);
    if (!sourceFile)
        return prior;
    // Find the enclosing function
    const node = findNodeAtPosition(sourceFile, position);
    if (!node)
        return prior;
    const enclosingFunction = findEnclosingFunction(node);
    if (!enclosingFunction)
        return prior;
    const directive = (0, utils_1.getDirective)(enclosingFunction, sourceFile, ts);
    // If we're in a workflow function, add workflow hooks to completions
    if (directive === 'use workflow') {
        const workflowCompletions = utils_1.WORKFLOW_HOOKS.map((hookName) => ({
            name: hookName,
            kind: ts.ScriptElementKind.functionElement,
            kindModifiers: '',
            sortText: '0', // Sort to top
            insertText: hookName,
            isRecommended: true,
        }));
        return {
            ...prior,
            entries: [...workflowCompletions, ...prior.entries],
        };
    }
    return prior;
}
function findNodeAtPosition(sourceFile, position) {
    function find(node) {
        if (position >= node.getStart() && position < node.getEnd()) {
            return tsserverlibrary_1.default.forEachChild(node, find) || node;
        }
        return undefined;
    }
    return find(sourceFile);
}
function findEnclosingFunction(node) {
    let current = node;
    while (current) {
        if (tsserverlibrary_1.default.isFunctionDeclaration(current) ||
            tsserverlibrary_1.default.isArrowFunction(current) ||
            tsserverlibrary_1.default.isFunctionExpression(current)) {
            return current;
        }
        current = current.parent;
    }
    return undefined;
}
