// @workflow/typescript-plugin/dist/hover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHoverInfo = getHoverInfo;
const utils_1 = require("./utils");
function getHoverInfo(fileName, position, program, ts) {
    const sourceFile = program.getSourceFile(fileName);
    if (!sourceFile) {
        return undefined;
    }
    // Find the node at the hover position
    let directiveNode;
    function visit(node) {
        if (ts.isStringLiteral(node) &&
            position >= node.getStart(sourceFile) &&
            position < node.getEnd()) {
            directiveNode = node;
            return;
        }
        ts.forEachChild(node, visit);
    }
    visit(sourceFile);
    if (!directiveNode) {
        return undefined;
    }
    const text = directiveNode.text;
    // Check if this is a directive
    if (text !== 'use workflow' && text !== 'use step') {
        return undefined;
    }
    // Check if this string is the first statement in a function body
    const parent = directiveNode.parent;
    if (!parent || !ts.isExpressionStatement(parent)) {
        return undefined;
    }
    const grandParent = parent.parent;
    if (!grandParent || !ts.isBlock(grandParent)) {
        return undefined;
    }
    const blockParent = grandParent.parent;
    if (!blockParent ||
        !(ts.isFunctionDeclaration(blockParent) ||
            ts.isArrowFunction(blockParent) ||
            ts.isFunctionExpression(blockParent))) {
        return undefined;
    }
    // Make sure this is the first statement
    if (grandParent.statements[0] !== parent) {
        return undefined;
    }
    // Get the parent function to determine directive type
    const directive = (0, utils_1.getDirective)(blockParent, sourceFile, ts);
    if (!directive) {
        return undefined;
    }
    const isWorkflow = directive === 'use workflow';
    const docUrl = isWorkflow
        ? 'https://useworkflow.dev/docs/foundations/workflows-and-steps#workflow-functions'
        : 'https://useworkflow.dev/docs/foundations/workflows-and-steps#step-functions';
    const directiveType = isWorkflow ? 'Workflow' : 'Step';
    return {
        kind: ts.ScriptElementKind.constElement,
        kindModifiers: '',
        textSpan: {
            start: directiveNode.getStart(sourceFile),
            length: directiveNode.getWidth(sourceFile),
        },
        displayParts: [
            {
                text: `Workflow DevKit - ${directiveType} Function`,
                kind: 'text',
            },
        ],
        documentation: [
            {
                text: `[Learn more](${docUrl}) about the \`"${directive}"\` directive.`,
                kind: 'text',
            },
        ],
    };
}
