// @workflow/builders/dist/standalone.js
import { BaseBuilder } from './base-builder.js';
export class StandaloneBuilder extends BaseBuilder {
    constructor(config) {
        super({
            ...config,
            dirs: ['.'],
        });
    }
    async build() {
        const inputFiles = await this.getInputFiles();
        const tsConfig = await this.getTsConfigOptions();
        const options = {
            inputFiles,
            tsBaseUrl: tsConfig.baseUrl,
            tsPaths: tsConfig.paths,
        };
        await this.buildStepsBundle(options);
        await this.buildWorkflowsBundle(options);
        await this.buildWebhookFunction();
        await this.createClientLibrary();
    }
    async buildStepsBundle({ inputFiles, tsPaths, tsBaseUrl, }) {
        console.log('Creating steps bundle at', this.config.stepsBundlePath);
        const stepsBundlePath = this.resolvePath(this.config.stepsBundlePath);
        await this.ensureDirectory(stepsBundlePath);
        await this.createStepsBundle({
            outfile: stepsBundlePath,
            inputFiles,
            tsBaseUrl,
            tsPaths,
        });
    }
    async buildWorkflowsBundle({ inputFiles, tsPaths, tsBaseUrl, }) {
        console.log('Creating workflows bundle at', this.config.workflowsBundlePath);
        const workflowBundlePath = this.resolvePath(this.config.workflowsBundlePath);
        await this.ensureDirectory(workflowBundlePath);
        await this.createWorkflowsBundle({
            outfile: workflowBundlePath,
            inputFiles,
            tsBaseUrl,
            tsPaths,
        });
    }
    async buildWebhookFunction() {
        console.log('Creating webhook bundle at', this.config.webhookBundlePath);
        const webhookBundlePath = this.resolvePath(this.config.webhookBundlePath);
        await this.ensureDirectory(webhookBundlePath);
        await this.createWebhookBundle({
            outfile: webhookBundlePath,
        });
    }
}
