// @workflow/nitro/dist/builders.js
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { BaseBuilder, createBaseBuilderConfig, VercelBuildOutputAPIBuilder, } from '@workflow/builders';
import { join } from 'pathe';
export class VercelBuilder extends VercelBuildOutputAPIBuilder {
    constructor(nitro) {
        super({
            ...createBaseBuilderConfig({
                workingDir: nitro.options.rootDir,
                dirs: ['.'], // Different apps that use nitro have different directories
            }),
            buildTarget: 'vercel-build-output-api',
        });
    }
    async build() {
        const configPath = join(this.config.workingDir, '.vercel/output/config.json');
        const originalConfig = JSON.parse(await readFile(configPath, 'utf-8'));
        await super.build();
        const newConfig = JSON.parse(await readFile(configPath, 'utf-8'));
        originalConfig.routes.unshift(...newConfig.routes);
        await writeFile(configPath, JSON.stringify(originalConfig, null, 2));
    }
}
export class LocalBuilder extends BaseBuilder {
    #outDir;
    constructor(nitro) {
        const outDir = join(nitro.options.buildDir, 'workflow');
        super({
            ...createBaseBuilderConfig({
                workingDir: nitro.options.rootDir,
                watch: nitro.options.dev,
                dirs: ['.'], // Different apps that use nitro have different directories
            }),
            buildTarget: 'next', // Placeholder, not actually used
        });
        this.#outDir = outDir;
    }
    async build() {
        const inputFiles = await this.getInputFiles();
        await mkdir(this.#outDir, { recursive: true });
        await this.createWorkflowsBundle({
            outfile: join(this.#outDir, 'workflows.mjs'),
            bundleFinalOutput: false,
            format: 'esm',
            inputFiles,
        });
        await this.createStepsBundle({
            outfile: join(this.#outDir, 'steps.mjs'),
            externalizeNonSteps: true,
            format: 'esm',
            inputFiles,
        });
        const webhookRouteFile = join(this.#outDir, 'webhook.mjs');
        await this.createWebhookBundle({
            outfile: webhookRouteFile,
            bundle: false,
        });
    }
}
