// @workflow/nitro/dist/vite.js
import { LocalBuilder } from './builders.js';
import nitroModule from './index.js';
import { workflowTransformPlugin } from '@workflow/rollup';
import { workflowHotUpdatePlugin } from '@workflow/vite';
import { createBuildQueue } from '@workflow/builders';
export function workflow(options) {
    let builder;
    const enqueue = createBuildQueue();
    return [
        workflowTransformPlugin(),
        {
            name: 'workflow:nitro',
            nitro: {
                setup: (nitro) => {
                    nitro.options.workflow = {
                        ...nitro.options.workflow,
                        ...options,
                        _vite: true,
                    };
                    if (nitro.options.dev) {
                        builder = new LocalBuilder(nitro);
                    }
                    return nitroModule.setup(nitro);
                },
            },
            // NOTE: This is a workaround because Nitro passes the 404 requests to the dev server to handle.
            // For workflow routes, we override to send an empty body to prevent Hono/Vite's SPA fallback.
            configureServer(server) {
                // Add middleware to intercept 404s on workflow routes before Vite's SPA fallback
                return () => {
                    server.middlewares.use((req, res, next) => {
                        // Only handle workflow webhook routes
                        if (!req.url?.startsWith('/.well-known/workflow/v1/')) {
                            return next();
                        }
                        // Wrap writeHead to ensure we send empty body for 404s
                        const originalWriteHead = res.writeHead;
                        res.writeHead = function (...args) {
                            const statusCode = typeof args[0] === 'number' ? args[0] : 200;
                            // NOTE: Workaround because Nitro passes 404 requests to the vite to handle.
                            // Causes `webhook route with invalid token` test to fail.
                            // For 404s on workflow routes, ensure we're sending the right headers
                            if (statusCode === 404) {
                                // Set content-length to 0 to prevent Vite from overriding
                                res.setHeader('Content-Length', '0');
                            }
                            // @ts-expect-error - Complex overload signature
                            return originalWriteHead.apply(this, args);
                        };
                        next();
                    });
                };
            },
        },
        workflowHotUpdatePlugin({
            builder: () => builder,
            enqueue,
        }),
    ];
}
