// @workflow/core/dist/observability.js
/**
 * Observability utilities for workflow inspection.
 * Shared between CLI and Web UI for consistent behavior.
 */
import { hydrateStepArguments, hydrateStepReturnValue, hydrateWorkflowArguments, hydrateWorkflowReturnValue, } from './serialization.js';
const STREAM_ID_PREFIX = 'strm_';
/**
 * Marker for stream reference objects that can be rendered as links
 */
export const STREAM_REF_TYPE = '__workflow_stream_ref__';
/**
 * Check if a value is a stream ID string
 */
export const isStreamId = (value) => {
    return typeof value === 'string' && value.startsWith(STREAM_ID_PREFIX);
};
/**
 * Check if a value is a StreamRef object
 */
export const isStreamRef = (value) => {
    return (value !== null &&
        typeof value === 'object' &&
        '__type' in value &&
        value.__type === STREAM_REF_TYPE &&
        'streamId' in value &&
        typeof value.streamId === 'string');
};
/**
 * Create a StreamRef object from a stream value.
 * This is used during hydration to convert serialized streams into
 * objects that can be rendered as links in the UI.
 */
const streamToStreamRef = (value) => {
    let streamId;
    if ('name' in value) {
        const name = String(value.name);
        if (!name.startsWith(STREAM_ID_PREFIX)) {
            streamId = `${STREAM_ID_PREFIX}${name}`;
        }
        else {
            streamId = name;
        }
    }
    else {
        streamId = `${STREAM_ID_PREFIX}null`;
    }
    return {
        __type: STREAM_REF_TYPE,
        streamId,
    };
};
const serializedStepFunctionToString = (value) => {
    if (!value)
        return 'null';
    if (typeof value !== 'object')
        return 'null';
    if ('stepId' in value) {
        const stepId = value.stepId;
        // TODO: Add closure vars to the string representation.
        // value.closureVars
        return `<step:${stepId}>`;
    }
    return '<function>';
};
/**
 * This is an extra reviver for devalue that takes any streams that would be converted,
 * into actual streams, and instead formats them as StreamRef objects for display in the UI.
 *
 * This is mainly because we don't want to open any streams that we aren't going to read from,
 * and so we can get the string ID/name, which the serializer stream doesn't provide.
 */
const streamPrintRevivers = {
    ReadableStream: streamToStreamRef,
    WritableStream: streamToStreamRef,
    TransformStream: streamToStreamRef,
    StepFunction: serializedStepFunctionToString,
};
const hydrateStepIO = (step) => {
    return {
        ...step,
        input: step.input && Array.isArray(step.input) && step.input.length
            ? hydrateStepArguments(step.input, [], step.runId, globalThis, streamPrintRevivers)
            : step.input,
        output: step.output
            ? hydrateStepReturnValue(step.output, globalThis, streamPrintRevivers)
            : step.output,
    };
};
const hydrateWorkflowIO = (workflow) => {
    return {
        ...workflow,
        input: workflow.input && Array.isArray(workflow.input) && workflow.input.length
            ? hydrateWorkflowArguments(workflow.input, globalThis, streamPrintRevivers)
            : workflow.input,
        output: workflow.output
            ? hydrateWorkflowReturnValue(workflow.output, [], workflow.runId, globalThis, streamPrintRevivers)
            : workflow.output,
    };
};
const hydrateEventData = (event) => {
    if (!event.eventData) {
        return event;
    }
    const eventData = { ...event.eventData };
    // Events can have various eventData with non-devalued keys.
    // So far, only eventData.result is devalued (though this may change),
    // so we need to hydrate it specifically.
    try {
        if ('result' in eventData && typeof eventData.result === 'object') {
            eventData.result = hydrateStepReturnValue(eventData.result, globalThis, streamPrintRevivers);
        }
    }
    catch (error) {
        console.error('Error hydrating event data', error);
    }
    return {
        ...event,
        eventData,
    };
};
const hydrateHookMetadata = (hook) => {
    return {
        ...hook,
        metadata: hook.metadata && 'runId' in hook
            ? hydrateStepArguments(hook.metadata, [], hook.runId, globalThis, streamPrintRevivers)
            : hook.metadata,
    };
};
export const hydrateResourceIO = (resource) => {
    if (!resource) {
        return resource;
    }
    let hydrated;
    if ('stepId' in resource) {
        hydrated = hydrateStepIO(resource);
    }
    else if ('hookId' in resource) {
        hydrated = hydrateHookMetadata(resource);
    }
    else if ('eventId' in resource) {
        hydrated = hydrateEventData(resource);
    }
    else {
        hydrated = hydrateWorkflowIO(resource);
    }
    if ('executionContext' in hydrated) {
        const { executionContext: _, ...rest } = hydrated;
        return rest;
    }
    return hydrated;
};
/**
 * Extract all stream IDs from a value (recursively traverses objects/arrays)
 */
export function extractStreamIds(obj) {
    const streamIds = [];
    function traverse(value) {
        if (isStreamId(value)) {
            streamIds.push(value);
        }
        else if (Array.isArray(value)) {
            for (const item of value) {
                traverse(item);
            }
        }
        else if (value && typeof value === 'object') {
            for (const val of Object.values(value)) {
                traverse(val);
            }
        }
    }
    traverse(obj);
    return Array.from(new Set(streamIds)); // Remove duplicates
}
/**
 * Truncate a string to a maximum length, adding ellipsis if needed
 */
export function truncateId(id, maxLength = 12) {
    if (id.length <= maxLength)
        return id;
    return `${id.slice(0, maxLength)}...`;
}
