// @workflow/core/dist/runtime/suspension-handler.js
import { waitUntil } from '@vercel/functions';
import { WorkflowAPIError } from '@workflow/errors';
import { dehydrateStepArguments } from '../serialization.js';
import * as Attribute from '../telemetry/semantic-conventions.js';
import { serializeTraceCarrier } from '../telemetry.js';
import { queueMessage } from './helpers.js';
/**
 * Processes a single hook by creating it in the database and event log.
 */
async function processHook({ queueItem, world, runId, global, }) {
    try {
        // Create hook in database
        const hookMetadata = typeof queueItem.metadata === 'undefined'
            ? undefined
            : dehydrateStepArguments(queueItem.metadata, global);
        await world.hooks.create(runId, {
            hookId: queueItem.correlationId,
            token: queueItem.token,
            metadata: hookMetadata,
        });
        // Create hook_created event in event log
        await world.events.create(runId, {
            eventType: 'hook_created',
            correlationId: queueItem.correlationId,
        });
    }
    catch (err) {
        if (WorkflowAPIError.is(err)) {
            if (err.status === 409) {
                // Hook already exists (duplicate hook_id constraint), so we can skip it
                console.warn(`Hook with correlation ID "${queueItem.correlationId}" already exists, skipping: ${err.message}`);
                return;
            }
            else if (err.status === 410) {
                // Workflow has already completed, so no-op
                console.warn(`Workflow run "${runId}" has already completed, skipping hook "${queueItem.correlationId}": ${err.message}`);
                return;
            }
        }
        throw err;
    }
}
/**
 * Processes a single step by creating it in the database and queueing execution.
 */
async function processStep({ queueItem, world, runId, workflowName, workflowStartedAt, global, }) {
    const ops = [];
    const dehydratedInput = dehydrateStepArguments({
        args: queueItem.args,
        closureVars: queueItem.closureVars,
    }, global);
    try {
        const step = await world.steps.create(runId, {
            stepId: queueItem.correlationId,
            stepName: queueItem.stepName,
            input: dehydratedInput,
        });
        waitUntil(Promise.all(ops).catch((opErr) => {
            // Ignore expected client disconnect errors (e.g., browser refresh during streaming)
            const isAbortError = opErr?.name === 'AbortError' || opErr?.name === 'ResponseAborted';
            if (!isAbortError)
                throw opErr;
        }));
        await queueMessage(world, `__wkf_step_${queueItem.stepName}`, {
            workflowName,
            workflowRunId: runId,
            workflowStartedAt,
            stepId: step.stepId,
            traceCarrier: await serializeTraceCarrier(),
            requestedAt: new Date(),
        }, {
            idempotencyKey: queueItem.correlationId,
        });
    }
    catch (err) {
        if (WorkflowAPIError.is(err) && err.status === 409) {
            // Step already exists, so we can skip it
            console.warn(`Step "${queueItem.stepName}" with correlation ID "${queueItem.correlationId}" already exists, skipping: ${err.message}`);
            return;
        }
        throw err;
    }
}
/**
 * Processes a single wait by creating the event and calculating timeout.
 * @returns The timeout in seconds, or null if the wait already exists.
 */
async function processWait({ queueItem, world, runId, }) {
    try {
        // Only create wait_created event if it hasn't been created yet
        if (!queueItem.hasCreatedEvent) {
            await world.events.create(runId, {
                eventType: 'wait_created',
                correlationId: queueItem.correlationId,
                eventData: {
                    resumeAt: queueItem.resumeAt,
                },
            });
        }
        // Calculate how long to wait before resuming
        const now = Date.now();
        const resumeAtMs = queueItem.resumeAt.getTime();
        const delayMs = Math.max(1000, resumeAtMs - now);
        return Math.ceil(delayMs / 1000);
    }
    catch (err) {
        if (WorkflowAPIError.is(err) && err.status === 409) {
            // Wait already exists, so we can skip it
            console.warn(`Wait with correlation ID "${queueItem.correlationId}" already exists, skipping: ${err.message}`);
            return null;
        }
        throw err;
    }
}
/**
 * Handles a workflow suspension by processing all pending operations (hooks, steps, waits).
 * Hooks are processed first to prevent race conditions, then steps and waits in parallel.
 */
export async function handleSuspension({ suspension, world, runId, workflowName, workflowStartedAt, span, }) {
    // Separate queue items by type for parallel processing
    const stepItems = suspension.steps.filter((item) => item.type === 'step');
    const hookItems = suspension.steps.filter((item) => item.type === 'hook');
    const waitItems = suspension.steps.filter((item) => item.type === 'wait');
    // Process all hooks first to prevent race conditions
    await Promise.all(hookItems.map((queueItem) => processHook({
        queueItem,
        world,
        runId,
        global: suspension.globalThis,
    })));
    // Then process steps and waits in parallel
    const [, waitTimeouts] = await Promise.all([
        Promise.all(stepItems.map((queueItem) => processStep({
            queueItem,
            world,
            runId,
            workflowName,
            workflowStartedAt,
            global: suspension.globalThis,
        }))),
        Promise.all(waitItems.map((queueItem) => processWait({
            queueItem,
            world,
            runId,
        }))),
    ]);
    // Find minimum timeout from waits
    const minTimeoutSeconds = waitTimeouts.reduce((min, timeout) => {
        if (timeout === null)
            return min;
        if (min === null)
            return timeout;
        return Math.min(min, timeout);
    }, null);
    span?.setAttributes({
        ...Attribute.WorkflowRunStatus('workflow_suspended'),
        ...Attribute.WorkflowStepsCreated(stepItems.length),
        ...Attribute.WorkflowHooksCreated(hookItems.length),
        ...Attribute.WorkflowWaitsCreated(waitItems.length),
    });
    // If we encountered any waits, return the minimum timeout
    if (minTimeoutSeconds !== null) {
        return { timeoutSeconds: minTimeoutSeconds };
    }
    return {};
}
